Allow TensorMethods.h to include Dispatcher.h (alternative)

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24253 [test] Benchmarks
* #23695 [wip] Fixed size arrays
* #23668 [wip] Call aten ops through c10 dispatcher
* #23667 Register ATen ops with c10
* **#23888 Allow TensorMethods.h to include Dispatcher.h (alternative)**

This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Allow TensorMethods.h to include Dispatcher.h (alternative)

This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

dzhulgakov commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -1,823 +1,4 @@
-
-} // namespace at
-
+#include <ATen/core/TensorBody.h>
```
I'd add a comment explaining why we have this setup

dzhulgakov reviewed (commented):
many of the builds are broken (though I didn't looks why)

otherwise looks legit

ezyang commented:
deferring to @dzhulgakov

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

dzhulgakov commented:
Looks good. There's some obscure error on windows build: `00:20:49 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\ATen/core/dispatch/OperatorOptions.h(15): error C2059: syntax error: '='` Otherwise I think it's good.

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

dzhulgakov reviewed (approved):
Pre-stamping assuming that builds pass

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer committed:
Update on "Allow TensorMethods.h to include Dispatcher.h (alternative)"


This is an alternative to https://github.com/pytorch/pytorch/pull/23684.

Instead of splitting a bunch of headers into declaration and definition, we change tensor includes to only include the tensor declaration when the tensor definition isn't needed.

Differential Revision: [D16673569](https://our.internmc.facebook.com/intern/diff/D16673569/)

smessmer commented:
The failing windows build is flaky, see https://github.com/pytorch/pytorch/issues/25393

Evidence: Three builds back in commit https://github.com/pytorch/pytorch/pull/23888/commits/b1e26811a262f3f05a32f4315e8968e545df7c78, the Windows build passed but then rocm had some issues.
